server/src/main/java/org/opensearch/action/admin/cluster/remotestore/restore/RemoteStoreRestoreClusterStateListener.java

```diff
@@ -0,0 +1,89 @@
+import org.opensearch.snapshots.RestoreService;
+
+import static org.opensearch.snapshots.RestoreService.restoreInProgress;
+
+/**
+ * Transport listener for cluster state updates
+ *
+ * @opensearch.internal
+ */
+public class RemoteStoreRestoreClusterStateListener implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(RemoteStoreRestoreClusterStateListener.class);
+
+    private final ClusterService clusterService;
+    private final String uuid;
+    private final ActionListener<RestoreRemoteStoreResponse> listener;
+
+    private RemoteStoreRestoreClusterStateListener(
+        ClusterService clusterService,
+        RestoreService.RestoreCompletionResponse response,
+        ActionListener<RestoreRemoteStoreResponse> listener
+    ) {
+        this.clusterService = clusterService;
+        this.uuid = response.getUuid();
+        this.listener = listener;
+
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent changedEvent) {
+        final RestoreInProgress.Entry prevEntry = restoreInProgress(changedEvent.previousState(), uuid);
+        final RestoreInProgress.Entry newEntry = restoreInProgress(changedEvent.state(), uuid);
+        if (prevEntry == null) {
+            // When there is a cluster-manager failure after a restore has been started, this listener might not be registered
+            // on the current cluster-manager and as such it might miss some intermediary cluster states due to batching.
+            // Clean up listener in that case and acknowledge completion of restore operation to client.
+            clusterService.removeListener(this);
+            listener.onResponse(new RestoreRemoteStoreResponse((RestoreInfo) null));
+        } else if (newEntry == null) {
+            clusterService.removeListener(this);
+            ImmutableOpenMap<ShardId, RestoreInProgress.ShardRestoreStatus> shards = prevEntry.shards();
+            assert prevEntry.state().completed() : "expected completed remote store restore state but was " + prevEntry.state();
+            assert RestoreService.completed(shards) : "expected all restore entries to be completed";
+            RestoreInfo ri = new RestoreInfo(
+                "remote_store",
+                prevEntry.indices(),
+                shards.size(),
+                +shards.size() - RestoreService.failedShards(shards)
+            );
+            RestoreRemoteStoreResponse response = new RestoreRemoteStoreResponse(ri);
+            logger.debug("restore from remote store completed");
+            listener.onResponse(response);
+        } else {
+            // restore not completed yet, wait for next cluster state update
+        }
+    }
+
+    /**
+     * Creates a cluster state listener and registers it with the cluster service. The listener passed as a
+     * parameter will be called when the restore is complete.
+     */
+    public static void createAndRegisterListener(
+        ClusterService clusterService,
```

Comment:
Why do we need this proxy method which does nothing but call `clusterService::addListener`?

Reply:
Ah ... copy/paste from `RestoreClusterStateListener` .....

Reply:
Yes, most of the classes related to `_remotestore/_restore` API are used as it is from the corresponding snapshot implementation (with very minor changes, mostly related to naming)

Reply:
I propose we create a generic `RestoreClusterStateListener<T>` if its just a copy/paste use to ensure code becomes maintainable

Reply:
These implementations may not be same going forward. Today, we are using most of the constructs that were used in `_snapshot/_restore` but the `_remotestore/_restore` API will evolve in the future.

Reply:
We can always extend and reuse functionality and still add more features going forward. Not a fan of maintaining redundant code. It's also fine if you want to do a follow up